src/pkcs15init/pkcs15-openpgp.c

```diff
@@ -419,29 +419,42 @@ static int openpgp_store_data(struct sc_pkcs15_card *p15card, struct sc_profile
 	case SC_PKCS15_TYPE_CERT:
 		cinfo = (sc_pkcs15_cert_info_t *) obj->data;
 		cid = &(cinfo->id);
+		unsigned int tag = 0x7F21;
 
 		if (cid->len != 1) {
 			sc_log(card->ctx, "ID=%s is not valid.", sc_dump_hex(cid->value, cid->len));
 			LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
 		}
 
 		/* OpenPGP card v.2 contains only 1 certificate */
-		if (cid->value[0] != 3) {
+		if (cid->value[0] != 3 && p15card->card->type < SC_CARD_TYPE_OPENPGP_V3) {
 			sc_log(card->ctx,
 			       "This version does not support certificate ID = %d (only ID=3 is supported).",
 			       cid->value[0]);
 			LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
 		}
+
+
 		/* Just update the certificate DO */
+		u8 param = (u8) (2 - (cid->value[0] - 1));
```

Comment:
A check for `SC_CARD_TYPE_OPENPGP_V3` is needed here as `SC_CARDCTL_OPENPGP_SELECT_DATA` does not exist in older cards (import fails indeed, I checked it).

You may put it in an else block above.

Reply:
Isn't this already enforced by the if-statement before:
```C
/* OpenPGP card v.2 contains only 1 certificate */
if (cid->value[0] != 3 && p15card->card->type < SC_CARD_TYPE_OPENPGP_V3) {
    sc_log(card->ctx,
        "This version does not support certificate ID = %d (only ID=3 is supported).",
        cid->value[0]);
    LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
}
```
this should ensure the following lines won't be reach w/o a valid `card->type` and `cid->value[0]` combination, or am I misunderstanding this?

Reply:
The following code expects the full command range, but older cards do not have `SC_CARDCTL_OPENPGP_SELECT_DATA`. Thus, the following code fails for older cards (see line 440), no matter if they fullfill the `cid->value[0] == 3` in the mentioned if clause. That is to say, you have to differentiate your code for both types or change the code so that `SC_CARDCTL_OPENPGP_SELECT_DATA` is not used for key id 3 keys.

Do you see, what I mean?

Reply:
done by ensuring SELECT DATA is not called for < v3 openpgp cards